[PHP 8] Add a custom round method to workaround an incompatibility of the buil-in function

Konamiman opened the pull request:
### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

There's a number of places in the WooCommerce codebase where the built-in function `round` is executed passing a non-numeric value (not a number and not a string that can be parsed as a number), for example `round('')` when passing the value of a setting that has not been initialized. In PHP 7 this yields a value of 0, but in PHP 8 this throws an error.

This commit adds a `NumberUtil` class with a static `round` method, this method checks if the passed value is numeric and if so it just executes the built-in function right away, otherwise it applies `floatval` to the value before passing it to `round`. And all the calls to 'round' in the codebase are replaced with `NumberUtil::round`. 

### How to test the changes in this Pull Request:

Try to create a new order from the admin area.

* In PHP 7, it works with and without this change.
* In PHP 8, it will throw an error, which is fixed with this change.

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Dev - Add a NumberUtil::round method to workaround a breaking change in the buil-in round function in PHP8.

Konamiman committed:
Fix code style/sniffer errors in a few files.

Konamiman committed:
Add a 'NumberUtil' class with a 'round' method.

There's a number of places in the WooCommerce codebase where the
built-in function 'round' is executed passing a non-numeric value
(not a number and not a string that can be parsed as a number),
for example round(''). In PHP 7 this yields a value of 0, but in
PHP 8 this throws an error.

This commit adds a 'NumberUtil' class with a static 'round' method,
this method checks if the passed value is numeric and if so it just
executes the built-in function, otherwise it returns 0. And all the
calls to 'round' in the codebase are replaced with 'NumberUtil::round'.

claudiosanches commented on src/Utilities/NumberUtil.php:
```diff
@@ -0,0 +1,34 @@
+/**
+ * A class of utilities for dealing with numbers.
+ */
+final class NumberUtil {
```
Why not just `Number` since it's under the `Utilities` namespace?

Konamiman replied:
I think we already discussed this at some point. It's better to have explicit class names for readability, and the name `Number` seems to suggest that the class represents a number, which isn't the case. Also `StringUtil` already exists, and in that particular case `String` couldn't be used because it's a PHP reserved word.

vedanshujain commented on src/Utilities/NumberUtil.php:
```diff
@@ -0,0 +1,34 @@
+	 * @return float The value rounded to the given precision as a float, or the supplied default value.
+	 */
+	public static function round( $val, int $precision = 0, int $mode = PHP_ROUND_HALF_UP ) : float {
+		if ( is_numeric( $val ) ) {
```
there is a backward incompatibility here in that `is_numeric` will also reject strings when there are spaces or characters around a valid numeric value. See in PHP 7.2:
```sh
>>> $number_with_spaces = ' 1.5 ';
=> " 1.5 "
>>> NumberUtil::round( $number_with_spaces )
=> 0.0
>>> round( $number_with_spaces )
=> 2.0
>>>
```

perhaps using something like `floatval` before passing to round will be better? wdyt?

Konamiman replied:
Good point. I'll give that a try, thanks.

Konamiman replied:
@vedanshujain I've fixed it, now numbers with spaces work as expected. Also fixed a couple more of cases to act like the built-in `round`:

- For the boolean value `true` it now returns 1, not 0.
- For objects it now throws an error (that was the case in PHP 7 already), instead of returning 0.

vedanshujain replied:
Looks good. Thanks!

vedanshujain reviewed (changes requested):
Left a comment, other than that looks good!

Konamiman committed:
Fix a couple of incompatibilities with NumberUtils::round

- Passing a string that represents a number but has spaces (e.g. ' 1 ')
  now works as expected (the number is properly interpreted)
- Passing the boolean true now returns 1, not 0
- Passing an object throws an error, instead of returning 0

vedanshujain reviewed (approved):
LGTM!
